library/ns1_zone.py

```diff
@@ -110,173 +140,323 @@
+        zone based on Ansible parameters.
 
-        # zone found
+        :return: Results of module execution
+        :rtype: dict
+        """
+        changed = False
+        state = self.module.params.get("state")
+        zone = self.get_zone(self.module.params.get("name"))
+        if state == "present":
+            changed, zone = self.present(zone)
+        if state == "absent":
+            changed = self.absent(zone)
+        return self.build_result(changed, zone)
+
+    def present(self, zone):
+        """Handles use case where desired state of zone is present.
+        If zone is provided, it is updated with params from Ansible that
+        differ from existing values. If zone is not provided, a new zone will
+        be created with params from Ansible.
+
+        :param zone: Zone object of existing zone returned by NS1
+        :type zone: dict, optional
+        :return: Tuple in which first value reflects whether or not a change
+        occured and second value is new or updated zone object
+        :rtype: tuple(bool, dict)
+        """
+        changed = False
+        want = self.sanitize_params(self.module.params)
         if zone:
-            if state == "absent":
-                if self.module.check_mode:
-                    # short circut in check mode
-                    self.module.exit_json(changed=True)
-
-                zone.delete(errback=self.errback_generator())
-                self.module.exit_json(changed=True)
-            else:
-                self.update(zone)
+            """
+            diff = self.compare_params(zone.data, want)
+            if diff:
+                changed = True
+                zone = self.update(zone, diff)
+            """
+            changed, zone = self.update_on_diff(zone, want)
         else:
-            if state == "absent":
-                self.module.exit_json(changed=False)
-            else:
-                if self.module.check_mode:
-                    # short circuit in check mode
-                    self.module.exit_json(changed=True)
-                zone = self.ns1.createZone(
-                    self.module.params.get('name'),
-                    errback=self.errback_generator(),
-                    **self.api_params()
-                )
-                self.module.exit_json(
-                    changed=True, id=zone['id'], data=zone.data)
+            changed = True
+            zone = self.create(want)
+        return changed, zone
+
+    def update_on_diff(self, zone, want):
+        """triggers update of zone if diff between zone and desired state in want
+
+        :param zone: Zone object of existing zone returned by NS1
+        :type zone: dict
+        :param want: Desired state of zone
+        :type want: dict
+        :return: Tuple in which first value reflects whether or not a change
+        occured and second value is new or updated zone object
+        :rtype: tuple(bool, dict)
+        """
+        changed = False
```

Comment:
this `changed` stuff seems verbose to me, seems like this could just be 
```
if diff:
    return True, self.update(zone, diff)
return False, zone
```
But maybe that is not the Python way

Reply:
This is partly a holdover from the way this module was previously structured and partly a personal preference thing. My Python always tends to favor Single-Entry Single-Exit (as opposed to Go, where this would be tough to follow). It's def more verbose, but makes debugging easier imo.

In any case, I'm with you that it's probably unnecessary in a small funcs like these.

Reply:
Added multiple returns